src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -3115,7 +3115,19 @@ static QVariant fcnFormatNumber( const QVariantList &values, const QgsExpression
     parent->setEvalErrorString( QObject::tr( "Number of places must be positive" ) );
     return QVariant();
   }
-  return QStringLiteral( "%L1" ).arg( value, 0, 'f', places );
```

Comment:
Would the normal `QLocale::toString()` also be an option?
I'm wondering how to give the user control over "portable expressions", so he has a predictable result wherever it's executed.

One idea would be to put a `QLocale` in the context, defaulting to the system locale and then have a `with_locale` method on top that works like the `with_variable`, e.g.

```
with_locale('it', format_number(1999.567,2))
```

Reply:
OK, re predictability, we could hardcode the decimal point and thousand separator to "." and "," for format_number(). If someone wants a localized version after that, he can use a string replace function, or a new locale function could be added (full disclosure: not by me at this moment, I need a functional format_number() function for an ongoing project, no time for more :wink:).

Reply:
@m-kuhn I like your idea. I think we should always stick to the concept of the default locale as set in the user settings, and allow the user to override when needed. Mixing the two approaches would lead to (a greater) confusion.

Reply:
@m-kuhn , I can't simply use QLocale().toString(...) as I'll end up hitting the very issue this PR tries to address, which is that the QLocale used in QGIS now disables thousand separators.

That said, you gave me an idea for an alternative, easier approach. Coming in a minute :)

Reply:
@nirvn  maybe with a few more LOC you can clone default locale, change the decimal separator setting and use the cloned locale for .toString

Reply:
@elpaso , yeah, that's the alternative approach I'm compiling now.

```
  QLocale locale = QLocale();
  locale.setNumberOptions( locale.numberOptions() &= ~QLocale::NumberOption::OmitGroupSeparator );
  return locale.toString( value, 'f', places );
```

Reply:
What's the cost of creating a locale? Maybe this should be stored within the current QgsExpressionContext to avoid re-creating for every feature evaluated.

Reply:
@nyalldawson , in Python, it took 1.3sec to copy QLocale() and set the number options 1 million times:

```
for x in range(1000000):
    locale = QLocale()
    locale.setNumberOptions(locale.numberOptions() & QLocale.OmitGroupSeparator)
```

Reply:
Ok, no issue there then!